### PR TITLE
Add a CMake option to enable IPO

### PR DESCRIPTION
**Pull request by Kaffeine**
The main motivation is to reduce the size of binaries for release build (following #5074).
The feature is not enabled by default because it increases the build time and the effect should be valued case-by-case.

### Linux
Linux build time changed from 2:01 to 2:56 (with GNU gold) or 3:20 (default ld).

| Executable | Before  | After |
| ------------- | ---------|---- |
| config_retrieve | 886K | 43K |
| config_store | 890K | 51K |
| DDNet | 4.1M | 4.1M |
| DDNet-Server | 1.6M | 1.6M |
| dilate | 898K | 35K |
| map_convert_07 | 902K | 59K |
| map_diff | 890K | 47K  |
| map_extract | 898K | 47K  |

### macOS
macOS build time remained to be around 2:30.

The size of `.dmg` image without IPO is `47.6Mb`, the size with IPO enabled is `46.4Mb` which is 1.2Mb lighter. I didn't check the internals.

### Windows
Windows build time changed from 3:10 to 5:05

The binaries became even _bigger_ with IPO enabled. Nonsense, maybe IPO is messed up with some other project-specific compiler settings. I didn't investigate.

## Conclusion
It makes sense to (manually) enable IPO for macOS and Linux builds for production builds (probably update https://github.com/ddnet/ddnet-scripts/blob/master/release/build.sh for that).

### CODE REVIEW

**Comment by Jupeyy** [comment]
Do we really need to change code for this?

Just use `cmake .. -DCMAKE_INTERPROCEDURAL_OPTIMIZATION=ON`

We don't really support very old compilers anyway

**Comment by Kaffeine** [comment]
> Do we really need to change code for this?
> Just use `cmake .. -DCMAKE_INTERPROCEDURAL_OPTIMIZATION=ON`

I tried that at first. Globally enabled IPO fails on GTEST (warning about CMP0069 treated as error).

> Also why gold linker? xd isn't mold the currently fastest linker?

I've listed the numbers in PR. GA build for Linux without IPO: 2:01 minutes, with IPO: 3:20 (default ld), with IPO *and* GNU gold: 2:56.

**Comment by Jupeyy** [comment]
> I tried that at first. Globally enabled IPO fails on GTEST (warning about CMP0069 treated as error).

Ok cannot reproduce this works without any problems for me, but maybe it happens if you download GTEST over the DOWNLOAD_GTEST option.

since we had a pr with LTO before #4614 
@def- @heinrich5991 
What do you think about LTO for production use, in this case the argument is about smaller files. A few MBytes. Worth it or not? Is the download size the important factor? E.g. how much difference is there in the zip file.
Do you think we should add extra code for it, instead of just relying on CMake?

I guess it won't really hurt. Can MinGW already do LTO? :D

**Comment by Kaffeine** [comment]
> > I tried that at first. Globally enabled IPO fails on GTEST (warning about CMP0069 treated as error).
> 
> Ok cannot reproduce this works without any problems for me, but maybe it happens if you download GTEST over the DOWNLOAD_GTEST option.
> 
This is the (failed) build with https://github.com/infclass/infclass-client/actions/runs/2285963161 `CMAKE_INTERPROCEDURAL_OPTIMIZATION=ON`

> Do you think we should add extra code for it, instead of just relying on CMake?

... extra *CMake* code, instead of just relying on CMake *global property*.

**Comment by Jupeyy** [comment]
> This is the (failed) build with https://github.com/infclass/infclass-client/actions/runs/2285963161 `CMAKE_INTERPROCEDURAL_OPTIMIZATION=ON`

Mh ok, sounds to me like GTest should not be affected by whatever compile flag we pass to DDNet's build system tho.
Would be the cleaner fix imo

**Comment by Kaffeine** [comment]
If you don't want to save a few MB on macOS and Linux then I'm fine with closing this until better times. Not a big deal :smiley: 
(In that case I'll add https://github.com/ddnet/ddnet/pull/5076/commits/f595318b830653ca13f462e6949fbf80eb7b6860 to some other PR)

**Comment by Jupeyy** [comment]
Im in favor of adding LTO to the Release. Just prefer to keep stuff as simple as possible.

So thanks for firing up this discussion and also share what benefit it gives

**Comment by def-** [comment]
Yes, LTO for Release builds sounds fine to me, it had no noticable performance change, but saving a few MB in executable size is fine.

**Comment by Kaffeine** [comment]
For Release builds (done everytime by GitHub Action) or for the production builds (via https://github.com/ddnet/ddnet-scripts/blob/master/release/build.sh)?

I mean: should I add a commit which enables LTO in macOS/Linux CI? Maybe I can also try to use mold for CI, if you're interesting.

**Comment by def-** [comment]
mold is a bit more complicated to set up right now, probably linking time doesn't matter that much.

**Comment by Learath2** [comment]
Definitely for using this in Release configuration. I'm curious what went wrong with Windows, so let's create a tracking issue after we merge this so we can look into it.

**Comment by Kaffeine** [comment]
Hm, I thought that https://github.com/ddnet/ddnet/pull/5076/commits/c281414ecdaa75387dc7726383c95f79470025d3 won't change anything, but in fact CMake version for 20.04 changes from 3.16.3 to 3.23.1.

Okay, I dropped that commit so we can test the build with CMake 3.16.

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -106,7 +109,7 @@ jobs:
         mkdir debug
         cd debug
         ${{ matrix.cmake-path }}cmake --version
-        ${{ matrix.cmake-path }}cmake ${{ matrix.cmake-args }} -DCMAKE_BUILD_TYPE=Debug -Werror=dev -DDOWNLOAD_GTEST=ON -DDEV=ON -DCMAKE_RUNTIME_OUTPUT_DIRECTORY_DEBUG=. ..
+        ${{ matrix.cmake-path }}cmake ${{ matrix.cmake-args }} ${{ matrix.cmake-release-args }} -DCMAKE_BUILD_TYPE=Debug -Werror=dev -DDOWNLOAD_GTEST=ON -DDEV=ON -DCMAKE_RUNTIME_OUTPUT_DIRECTORY_DEBUG=. ..
```
Comment (def-) [inline .github/workflows/build.yaml:112]: This seems wrong. Why enable -DIPO=ON in Debug builds? There is a release build below.

File: CMakeLists.txt [hunk 2]
```diff
@@ -126,6 +126,7 @@ option(DEV "Don't generate stuff necessary for packaging" OFF)
 option(VULKAN "Enable the vulkan backend" ${AUTO_VULKAN_BACKEND})
 
 option(EXCEPTION_HANDLING "Enable exception handling (only works with Windows as of now)" OFF)
+option(IPO "Enable interprocedural optimizations" OFF)
```
Comment (def-) [inline CMakeLists.txt:129]: Maybe we should default IPO to ON in Release builds.

**Review by def- — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yaml (hunk 1 above)
Comment (Kaffeine) [reply .github/workflows/build.yaml:112]: Oh.

File: CMakeLists.txt (hunk 2 above)
Comment (Kaffeine) [reply CMakeLists.txt:129]: Ok :+1:

Comment (Kaffeine) [reply CMakeLists.txt:129]: Only for Linux and macOS?

Comment (def-) [reply CMakeLists.txt]: How did you build on Windows? Is it with Visual Studio? If so, let's turn it off by default on Visual Studio with a comment explaining why.

Comment (Kaffeine) [reply CMakeLists.txt]: Locally I have only Linux. I use CI to build for any other platform.

> If so, let's turn it off by default on Visual Studio

Hm, maybe IPO will be beneficial for mingw builds indeed but we don't have it in CI.

> with a comment explaining why

Okay.

File: CMakeLists.txt [hunk 3]
```diff
@@ -95,9 +95,11 @@ else()
 endif()
 
 set(AUTO_DEPENDENCIES_DEFAULT OFF)
+set(AUTO_IPO ON)
 set(AUTO_VULKAN_BACKEND ON)
 if(TARGET_OS STREQUAL "windows")
   set(AUTO_DEPENDENCIES_DEFAULT ON)
+  set(AUTO_IPO OFF)
```
Comment (def-) [inline CMakeLists.txt:102]: `if(MSVC)`
CI uses MSVC, I will attempt it with MinGW when the change is ready.

**Review by def- — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 4]
```diff
@@ -95,9 +95,11 @@ else()
 endif()
 
 set(AUTO_DEPENDENCIES_DEFAULT OFF)
+set(AUTO_IPO ON)
```
Comment (def-) [inline CMakeLists.txt:98]: Wait, only set it to ON in Release builds, not in Debug.

Comment (Kaffeine) [reply CMakeLists.txt]: Yeah, I've noticed the missing condition right after push.

**Comment by def-** [comment]
bors r+
Thanks

**Review by def- — APPROVED** [review]

**Pull request merged**